subprojects/maven/src/main/java/org/gradle/api/publish/maven/internal/publication/MavenComponentParser.java

```diff
@@ -96,21 +102,22 @@ public class MavenComponentParser {
     private final VersionRangeMapper versionRangeMapper;
     private final DocumentationRegistry documentationRegistry;
     private final NotationParser<Object, MavenArtifact> mavenArtifactParser;
-    private final VariantDependencyResolverFactory variantDependencyResolverFactory;
+    private final DependencyCoordinateResolverFactory dependencyCoordinateResolverFactory;
 
+    @Inject
     public MavenComponentParser(
+        ObjectFactory objectFactory,
         PlatformSupport platformSupport,
         VersionRangeMapper versionRangeMapper,
         DocumentationRegistry documentationRegistry,
         VersionMappingStrategyInternal versionMappingStrategy,
-        ProjectDependencyPublicationResolver projectDependencyResolver,
         NotationParser<Object, MavenArtifact> mavenArtifactParser
     ) {
         this.platformSupport = platformSupport;
         this.versionRangeMapper = versionRangeMapper;
         this.documentationRegistry = documentationRegistry;
         this.mavenArtifactParser = mavenArtifactParser;
-        this.variantDependencyResolverFactory = new DefaultVariantDependencyResolverFactory(projectDependencyResolver, versionMappingStrategy);
+        this.dependencyCoordinateResolverFactory = objectFactory.newInstance(DefaultDependencyCoordinateResolverFactory.class, versionMappingStrategy);
```

Comment:
🤔 could be a service?

Reply:
See https://github.com/gradle/gradle/pull/26356#discussion_r1331868452